src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+                ParseGroupingEmbeddedExpression(GetNewOptionsFromToken(_options, optionsToken)), ParseGroupingCloseParen());
+
+        private static bool IsTextChar(RegexToken currentToken, char ch)
+            => currentToken.Kind == RegexKind.TextToken && currentToken.VirtualChars.Length == 1 && currentToken.VirtualChars[0].Char == ch;
+
+        private static RegexOptions GetNewOptionsFromToken(RegexOptions currentOptions, RegexToken optionsToken)
+        {
+            var copy = currentOptions;
+            var on = true;
+            foreach (var ch in optionsToken.VirtualChars)
+            {
+                switch (ch.Char)
+                {
+                    case '-': on = false; break;
+                    case '+': on = true; break;
+                    default:
+                        var newOption = OptionFromCode(ch);
+                        if (on)
+                        {
+                            copy |= newOption;
+                        }
+                        else
+                        {
+                            copy &= ~newOption;
+                        }
+                        break;
+                }
+            }
+
+            return copy;
+        }
+
+        private static RegexOptions OptionFromCode(VirtualChar ch)
+        {
+            switch (ch)
+            {
+                case 'i': case 'I': return RegexOptions.IgnoreCase;
+                case 'm': case 'M': return RegexOptions.Multiline;
+                case 'n': case 'N': return RegexOptions.ExplicitCapture;
+                case 's': case 'S': return RegexOptions.Singleline;
+                case 'x': case 'X': return RegexOptions.IgnorePatternWhitespace;
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        private RegexBaseCharacterClassNode ParseCharacterClass()
+        {
+            var openBracketToken = _currentToken;
+            Debug.Assert(openBracketToken.Kind == RegexKind.OpenBracketToken);
+            var caretToken = CreateMissingToken(RegexKind.CaretToken);
+            var closeBracketToken = CreateMissingToken(RegexKind.CloseBracketToken);
+
+            // trivia is not allowed anywhere in a character class
+            ConsumeCurrentToken(allowTrivia: false);
+            if (_currentToken.Kind == RegexKind.CaretToken)
+            {
+                caretToken = _currentToken;
+            }
+            else
+            {
+                MoveBackBeforePreviousScan();
+            }
+
+            // trivia is not allowed anywhere in a character class
+            ConsumeCurrentToken(allowTrivia: false);
+
+            var contents = ArrayBuilder<RegexExpressionNode>.GetInstance();
+            while (_currentToken.Kind != RegexKind.EndOfFile)
+            {
+                Debug.Assert(_currentToken.VirtualChars.Length == 1);
+
+                if (_currentToken.Kind == RegexKind.CloseBracketToken && contents.Count > 0)
```

Comment:
The check on `contents.Count` is surprising to me. I would have assumed the closing bracket could never be part of the contents. So `[]]` would be an empty char class followed by a misplaced bracket.
Instead, it looks like that is parsed a char class with one character (the closing bracket) in it. Is that correct?

Reply:
Yes.  That is both correct and *intentional*.  It matches the .net regex parsing logic.  I will add comment to indicate this is intentional.  We also have tests for exactly this.